src/handler/lib/policy.c

```diff
@@ -133,59 +400,190 @@ static int get_derived_pubkey(policy_parser_state_t *state, int key_index, uint8
     return 0;
 }
 
-/**
- * Pushes a node onto the stack. Returns 0 on success, -1 if the stack is exhausted.
- */
-static int state_stack_push(policy_parser_state_t *state, policy_node_t *policy_node) {
-    ++state->node_stack_eos;
+static void update_output(policy_parser_state_t *state, const uint8_t *data, size_t data_len) {
+    policy_parser_node_state_t *node = &state->nodes[state->node_stack_eos];
 
-    if (state->node_stack_eos >= MAX_POLICY_DEPTH) {
-        return -1;
-    }
+    node->length += data_len;
 
-    policy_parser_node_state_t *node = &state->nodes[state->node_stack_eos];
-    cx_sha256_init(&state->hash_context);
-    node->policy_node = policy_node;
-    node->step = 0;
-    node->mode = MODE_OUT_HASH;
-    node->hash_context = &state->hash_context;
+    PRINTF("ADD TO HASH: ");
+    for (unsigned int i = 0; i < data_len; i++) {
+        PRINTF("%02X", data[i]);
+    }
+    PRINTF("\n");
 
-    return 0;
+    if (node->mode == MODE_OUT_BYTES) {
+        if (!buffer_write_bytes(node->out_buf, data, data_len)) {
+            node->flags |= PROCESSOR_FLAG_OUTPUT_OVERFLOW;
+        };
+    } else {
+        crypto_hash_update(&node->hash_context->header, data, data_len);
+    }
 }
 
-/**
- * Pops a node the stack. If the node is in HASH mode, computes the hash.
- * Returns 0 on success, -1 on error.
- */
-static int state_stack_pop(policy_parser_state_t *state) {
-    policy_parser_node_state_t *node = &state->nodes[state->node_stack_eos];
+static inline void update_output_u8(policy_parser_state_t *state, uint8_t data) {
+    update_output(state, &data, 1);
+}
 
-    if (state->node_stack_eos >= MAX_POLICY_DEPTH || state->node_stack_eos <= -1) {
-        return -1;
+// outputs the minimal push opcode for an unsigned 32bit integer
+static void update_output_push_u32(policy_parser_state_t *state, uint32_t n) {
+    if (n == 0) {
+        update_output_u8(state, OP_0);
+    } else if (n <= 16) {
+        update_output_u8(state, 0x50 + n);
```

Comment:
This is just [Bitcoin Script](https://en.bitcoin.it/wiki/Script) being weird. Pushing 0 is `OP_0`, pushing numbers from 1 to 16 has dedicated opcodes from `OP_1` to `OP_16` :shrug: 

Moreover, the encoding of numbers is redundant (multiple ways of representing the same number), and there are rules about exactly which encoding should be used (added to avoid `scriptSig` malleability issues), which is why there is this function.